Build your own Probot (#46)
JasonEtco/jasonet.co: src/pages/posts/build-your-own-probot.md

---
title: Build your own Probot
date: '2019-12-08'
spoiler: Let's see what makes Probot tick; the API design, the authentication mechanisms, and the little helpers that make it easy to use.
---

> A lot of people look at Probot and wonder how they can extend it - add custom routes, integrate it with other platforms, control it's startup. I firmly believe that too many features and options can make a framework unweildy, so rather than show where all of those things can fit in Probot itself, we're going to take a look at building our own Probot.

Reviewer: ```suggestion
A lot of people look at Probot and wonder how they can extend it - add custom routes, integrate it with other platforms, control its startup. I firmly believe that too many features and options can make a framework unwieldy, so rather than show where all of those things can fit in Probot itself, we're going to take a look at building our own Probot.
```

  // It's valid, we emit an event
  if (isValidWebhook) {
    events.emit('github-event', {
      event: eventName,
      payload: req.body
    })
    // We tell GitHub that everything worked great!
    return res.send(200)
  } else {
    // Couldn't validate the webhook, so we send an "Unauthorized" code
    return res.send(401)
  }
})
```

If your app is deployed to a public URL, GitHub can make an HTTP request; but if you're just working locally, you'll have to use something like `localtunnel` or `ngrok` to expose your local machine to the internet. We built [Smee.io](https://smee.io) specifically for Probot, to receive webhook events locally - it's its own topic, so if you're interested in a post on Smee [let me know](https://twitter.com/JasonEtco)!

Somewhere behind the scenes, Probot has registered the event handlers of your Probot app. A standard app looks like this:

```js
module.exports = app => {
  app.on('event', handler)
}
```

Where `app` is an instance of the `Application` class, which is (sort of) an extension of `EventEmitter! We'll look at that shortly - there are still a things that Probot does before running your code, but first...

### Why an EventEmitter API?

> To me, it's the inherent simplicity of "something happened, so now run this code." When an issue is opened on GitHub, run this code. This let's you focus on your app code, and not care what Probot is doing under the hood. Similar to something like Express's `app.get()`, but this let's you register multiple event handlers for the same event, and have an infinite list of events (since they're just strings).

Reviewer: ```suggestion
To me, it's the inherent simplicity of "something happened, so now run this code." When an issue is opened on GitHub, run this code. This lets you focus on your app code, and not care what Probot is doing under the hood. Similar to something like Express's `app.get()`, but this lets you register multiple event handlers for the same event, and have an infinite list of events (since they're just strings).
```

### Two ways for GitHub Apps to authenticate

If you're here, I'll guess that you're familiar with what GitHub Apps are. They're an "actor" on GitHub, and can authenticate in two distinct ways:

* As itself, as an app
* As an installation. When you install an app on an account (a user or an organization), that creates an installation. Each installation has a unique `id` (the aforementioned payload value), and we can use that to create an API access token.

Usually, when your Probot app calls `context.github...`, `github` is an authenticated `Octokit` client that uses an **installation token**. We can get one of those by creating an access token as the app and then using that to create an access token for an installation, given its ID.

Authenticating as the app is done by combining the `APP_ID` and `PRIVATE_KEY`, using JWTs:

```js
const jsonwebtoken = require('jsonwebtoken')
// A helper function for generated an access token using the App ID and Private Key
function getSignedJsonWebToken({ app_id, privateKey }) {
  const now = Math.floor(Date.now() / 1000)
  const payload = {
    iat: now, // Issued at time

>     exp: now + 60 * 10 - 30, // JWT expiration time (10 minute maximum, 30 seconds of leeway)

Reviewer: If you want to grant leeway, you should be adding 30 seconds, not taking it away. 🙂 

```suggestion
    exp: now + 60 * 10 + 30, // JWT expiration time (10 minute maximum, 30 seconds of leeway)
```

Author: Hm - you're making sense here, but this is directly copied from Probot. I'm in favor of leaving as-is even if its not correct, maybe with a disclaimer 🤔

Reviewer: I think perhaps the problem is just the use of the word "leeway". Looks like the subtraction is intentional per the original issue and PR:

- https://github.com/probot/probot/issues/942
- https://github.com/octokit/app.js/pull/36

Reviewer: It's more of a preventative safety net to avoid potentially running into "...the maximum expiration time of 10 minutes, after which the API will start returning a 401 error".

Author: I'll just remove the second part of the comment, that level of detail isn't really necessary here 👍

Reviewer: https://github.com/octokit/app.js/pull/62

Reviewer: Test

Reviewer: Another reply which will open the "Leave a review" entry field.

Reviewer: One more test. 🤔

      const authString = `token ${accessToken}`
      return new Octokit({ auth: authString })
    }
  
    // Otherwise, authenticate as the app using a JWT
    const token = this.githubToken || this.app.getSignedJsonWebToken()
    return new Octokit({ auth: `Bearer ${token}` })
  }
}
```

Along with some of the earlier code we looked at, we would use the `Application` class in code like this:

```js
const events = new EventEmitter()
const application = new Application()

// This is code similar to what you'd find in your Probot app's code
application.on('issues', async context => {
  return context.github.issues.createComment({ ... })
})

// We register an event in our internal EventEmitter, as a sort of middleware
// for our `Application` event emitter. When thats triggered, we create an
// authenticated GitHub client, and then emit an event using the webhook's event
// name, but on the Application instance.
events.on('github-event', ({ event, payload }) => {
  const github = await application.auth(payload.installation.id)
  const context = { github, payload }
  return application.emit(event, context)
})

app.post('/', ...)
```

> A common question I hear is "how do I access `context` outside of an event handler?" - hopefully, this has helped explain how `context` is created, and is dependant on the `installation.id` value of the payload. If you have another way to get an installation ID (possibly through [API endpoints for finding an ID](https://developer.github.com/v3/apps/#get-a-repository-installation)), then you can totally do that!

Reviewer: ```suggestion
A common question I hear is "how do I access `context` outside of an event handler?" - hopefully, this has helped explain how `context` is created, and is dependent on the `installation.id` value of the payload. If you have another way to get an installation ID (possibly through [API endpoints for finding an ID](https://developer.github.com/v3/apps/#get-a-repository-installation)), then you can totally do that!
```

```json
"scripts": {
  "start": "probot run ./index.js"
}
```

This tells Probot to `require()` the `./index.js` file, and run the exported function with an instance of `Application`. That's done through [`Probot#load`](https://github.com/probot/probot/blob/61c05ea76e9ea31fbdea15c99f7c1cc613321db5/src/index.ts#L179), and somewhere down the line we have code like this:

```ts
// From https://github.com/probot/probot/blob/master/src/resolver.ts
import { sync } from 'resolve'

export function resolve (appFnId: string) {
  // appFnId would be `./index.js`
  return require(sync(appFnId, { basedir: process.cwd() }))
}
```

We're loading the code via Node.js's `require()` functionality. Then, once we've got the function loaded up, we simply call it:

```js
const application = new Application()
const yourApp = probot.load('./index.js')
yourApp(application)
```

Calling the function then calls `application.on()`, which registers the event handlers!

While this obviously isn't all of Probot, it really is the core of it. We get an event through the `POST` endpoint, use information from it to create an authenticated GitHub client, and run your app's code.

## A very simple Probot framework

> Using all of the above information, I wanted to take a crack at a simplified version of Probot that, asidie from the GitHub authentication mechanisms, has nothing to do with GitHub. You can take this framework and apply it to any other service that sends webhooks.

Reviewer: ```suggestion
Using all of the above information, I wanted to take a crack at a simplified version of Probot that, aside from the GitHub authentication mechanisms, has nothing to do with GitHub. You can take this framework and apply it to any other service that sends webhooks.
```

### Two ways for GitHub Apps to authenticate

If you're here, I'll guess that you're familiar with what GitHub Apps are. They're an "actor" on GitHub, and can authenticate in two distinct ways:

* As itself, as an app
* As an installation. When you install an app on an account (a user or an organization), that creates an installation. Each installation has a unique `id` (the aforementioned payload value), and we can use that to create an API access token.

Usually, when your Probot app calls `context.github...`, `github` is an authenticated `Octokit` client that uses an **installation token**. We can get one of those by creating an access token as the app and then using that to create an access token for an installation, given its ID.

Authenticating as the app is done by combining the `APP_ID` and `PRIVATE_KEY`, using JWTs:

```js
const jsonwebtoken = require('jsonwebtoken')
// A helper function for generated an access token using the App ID and Private Key
function getSignedJsonWebToken({ app_id, privateKey }) {
  const now = Math.floor(Date.now() / 1000)
  const payload = {
    iat: now, // Issued at time

>     exp: now + 60 * 10 - 30, // JWT expiration time (10 minute maximum, 30 seconds of leeway)

Reviewer: ```suggestion
    exp: now + 60 * 10 - 30, // JWT expiration time (10 minute maximum, 30 second safeguard)
```
